weather-oracle/sources/weather.move

```diff
@@ -0,0 +1,377 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// This module defines a weather oracle that can post weather updates for major cities around the world.
+module oracle::weather {
+    use std::option::{Self, Option};
+    use std::string::{Self, String};
+
+    use sui::dynamic_object_field as dof;
+    use sui::object::{Self, UID};
+    use sui::package;
+    use sui::transfer::{Self};
+    use sui::tx_context::{Self, TxContext};
+
+    /// Define a capability for the admin of the oracle.
+    struct AdminCap has key, store { id: UID }
+
+    /// // Define a one-time witness to create the `Publisher` of the oracle.
+    struct WEATHER has drop {}
+
+    // Define a struct for the weather oracle
+    struct WeatherOracle has key {
+        id: UID,
+        /// The address of the oracle.
+        address: address,
+        /// The name of the oracle.
+        name: String,
+        /// The description of the oracle.
+        description: String,
+    }
+
+    // Define a struct for each city that the oracle covers
+    struct CityWeatherOracle has key, store {
+        id: UID,
+        geoname_id: u32, // The unique identifier of the city
+        name: String, // The name of the city
+        country: String, // The country of the city
+        latitude: u32, // The latitude of the city in degrees
+        positive_latitude: bool, // Whether the latitude is positive (north) or negative (south)
+        longitude: u32, // The longitude of the city in degrees
+        positive_longitude: bool, // Whether the longitude is positive (east) or negative (west)
+        weather_id: u16, // The weather condition code
+        temp: u32, // The temperature in kelvin
+        pressure: u32, // The atmospheric pressure in hPa
+        humidity: u8, // The humidity percentage
+        visibility: u16, // The visibility in meters
+        wind_speed: u16, // The wind speed in meters per second
+        wind_deg: u16, // The wind direction in degrees
+        wind_gust: Option<u16>, // The wind gust in meters per second (optional)
+        clouds: u8, // The cloudiness percentage
+        dt: u32 // The timestamp of the weather update in seconds since epoch
+    }
+
+    // Define a struct for a non-fungible token (NFT) that represents a weather conditions for a city.
+    struct WeatherNFT has key, store {
```

Comment:
Similar to the above - which of the parameters are important, and which are metadata? Perhaps, a VecMap would help?